src/tactic/unify_equations.lean

```diff
@@ -0,0 +1,407 @@
+    rhs_whnf ← whnf_ginductive rhs,
+    unify_homogeneous eque type type lhs rhs lhs_whnf rhs_whnf u
+  | (app (app (app (app (const `heq [u]) lhs_type) lhs) rhs_type) rhs) := do
+    lhs_whnf ← whnf_ginductive lhs,
+    rhs_whnf ← whnf_ginductive rhs,
+    unify_heterogeneous eque lhs_type rhs_type lhs rhs lhs_whnf rhs_whnf u
+  | _ := fail! "Expected {equ} to be an equation, but its type is\n{t}."
+  end
+
+/--
+Given a list of display names of local hypotheses that are (homogeneous or
+heterogeneous) equations, `unify_equations` performs first-order unification on
+each hypothesis in order. See `tactic.interactive.unify_equations` for an
+example and an explanation of what unification does.
+
+Returns true iff the goal has been solved during the unification process.
+
+Note: you must make sure that the input names are unique in the context.
+-/
+meta def unify_equations : list name → tactic bool
+| [] := pure ff
+| (h :: hs) := do
+  res ← unify_equation_once h,
+  match res with
+  | simplified hs' := unify_equations $ hs' ++ hs
+  | not_simplified := unify_equations hs
+  | goal_solved := pure tt
+  end
+
+
+namespace interactive
+
+open lean.parser
+
+/--
+`unify_equations eq₁ ... eqₙ` performs a form of first-order unification on the
+hypotheses `eqᵢ`. The `eqᵢ` must be homogeneous or heterogeneous equations.
+Unification means that the equations are simplified using various facts about
+constructors. For instance, consider this goal:
+
+```
+P : ∀ n, fin n → Prop
+n m : ℕ
+f : fin n
+g : fin m
+h₁ : n + 1 = m + 1
+h₂ : f == g
+h₃ : P n f
+⊢ P m g
+```
+
+After `unify_equations h₁ h₂`, we get
+
+```
+P : ∀ n, fin n → Prop
+n : ℕ
+f : fin n
+h₃ : P n f
+⊢ P n f
+```
+
+In the example, `unify_equations` uses the fact that every constructor is
+injective to conclude `n = m` from `h₁`. Then it replaces every `m` with `n` and
+moves on to `h₂`. The types of `f` and `g` are now equal, so the heterogeneous
+equation turns into a homogeneous one and `g` is replaced by `f`. Note that the
+equations are processed from left to right, so `unify_equations h₂ h₁` would not
+simplify as much.
+
+In general, `unify_equations` uses the following steps on each equation until
+none of them applies any more:
+
+- Constructor injectivity: if `nat.succ n = nat.succ m` then `n = m`.
+- Substitution: if `x = e` for some hypothesis `x`, then `x` is replaced by `e`
+  everywhere.
+- No-confusion: `nat.succ n = nat.zero` is a contradiction. If we have such an
+  equation, the goal is solved immediately.
+- Cycle elimination: `n = nat.succ n` is a contradiction.
+- Redundancy: if `t = u` but `t` and `u` are already definitionally equal, then
+  this equation is removed.
+- Downgrading of heterogeneous equations: if `t == u` but `t` and `u` have the
+  same type (up to definitional equality), then the equation is replaced by
+  `t = u`.
+-/
+meta def unify_equations (eqs : interactive.parse (many ident)) :
+  tactic unit :=
+tactic.unify_equations eqs *> skip
+
```

Comment:
Not sure about the tags, maybe there are more that can be added from the list [here](https://leanprover-community.github.io/mathlib_docs/tactics.html) (click on Filter by tags on the right).
```suggestion
add_tactic_doc
{ name := "unify_equations",
  category := doc_category.tactic,
  decl_names := [`tactic.interactive.unify_equations],
  tags := ["simplification"]
}

```

Reply:
I didn't find any other tags that seem super relevant.